### PR TITLE
Use static cache for UFMatch lookups to improve performance

### PR DESCRIPTION
**Pull request by colemanw**
Overview
----
Avoid excessive database hits by caching the UF/CID match records.

Before
----
More db lookups; slower

After
---
Fewer db lookups; faster

Notes
----
This is a supplement to https://github.com/civicrm/civicrm-core/pull/12576 which can potentially cause the `CRM_Core_BAO_UFMatch::getUFId()` to be called dozens of times per request. This caching ensures that we don't suffer a performance hit in doing so.

### CODE REVIEW

**Comment by eileenmcnaughton** [comment]
@colemanw ... nope

**Comment by colemanw** [comment]
Whoops, php really don't like it when you try to check a key for a nonexistent array. Now fixed.

**Comment by colemanw** [comment]
@civicrm-builder retest this please

**Comment by colemanw** [comment]
@eileenmcnaughton ... yep

File: CRM/Core/BAO/UFMatch.php [hunk 1]
```diff
@@ -51,10 +51,12 @@ public static function create($params) {
     }
     $dao = new CRM_Core_DAO_UFMatch();
     $dao->copyValues($params);
+    // Fixme: this function cannot update records
     if (!$dao->find(TRUE)) {
       $dao->save();
+      Civi::$statics[__CLASS__][$params['domain_id']][(int) $dao->contact_id] = (int) $dao->uf_id;
```
Comment (eileenmcnaughton) [inline CRM/Core/BAO/UFMatch.php]: you are mixing & matching contact_id & uf_id

contact_id

File: CRM/Core/BAO/UFMatch.php [hunk 2]
```diff
@@ -478,20 +483,29 @@ public static function deleteUser($ufID) {
...
+    }
     $ufmatch = new CRM_Core_DAO_UFMatch();
-
     $ufmatch->uf_id = $ufID;
-    $ufmatch->domain_id = CRM_Core_Config::domainID();
+    $ufmatch->domain_id = $domainId;
     if ($ufmatch->find(TRUE)) {
-      return (int ) $ufmatch->contact_id;
+      $contactId = (int) $ufmatch->contact_id;
+      Civi::$statics[__CLASS__][$domainId][$contactId] = (int) $ufID;
```
Comment (eileenmcnaughton) [inline CRM/Core/BAO/UFMatch.php]: ah no that's ok

**Comment by eileenmcnaughton** [comment]
I read through this & stepped through this. The getUFId function is called 4 times from the contact summary page so this eliminated 3 DB queries.

The getContactID is called when editing a user but no obvious query reduction.

I only tried this on a single domain but I feel like it's pretty safe having stepped through it in a debugger & I guess it can sneak through without a test

**Pull request merged**